hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/execution/bulkinsert/TestBulkInsertInternalPartitionerForRows.java

```diff
@@ -172,9 +173,14 @@ private void testBulkInsertInternalPartitioner(BulkInsertPartitioner partitioner
       return;
     }
     Dataset<Row> actualRecords = (Dataset<Row>) partitioner.repartitionRecords(rows, numPartitions);
-    assertEquals(
-        enforceNumOutputPartitions ? numPartitions : rows.rdd().getNumPartitions(),
-        actualRecords.rdd().getNumPartitions());
+    if (isGloballySorted) {
+      // For GLOBAL_SORT, `df.sort` may generate smaller number of partitions than the specified parallelism
+      assertTrue(actualRecords.rdd().getNumPartitions() <= numPartitions);
```

Comment:
would like to understand why this is the behavior. what's affecting the number of partitions? is it the size of data?

Reply:
This is happening because Global Sort Bulk Insert partitioner is using coalesce instead of repartition.